kafka-client/src/main/java/dev/responsive/kafka/clients/OffsetRecorder.java

```diff
@@ -0,0 +1,163 @@
+ */
+class OffsetRecorder {
+  private final Map<RecordingKey, Long> uncommitted = new HashMap<>();
+  private final Map<TopicPartition, Long> written = new HashMap<>();
+  private final ProducerListener producerListener = new ProducerListener();
+  private final ConsumerListener consumerListener = new ConsumerListener();
+  private final List<CommitCallback> commitCallback = new LinkedList<>();
+  private final boolean eos;
+
+  OffsetRecorder(final boolean eos) {
+    this.eos = eos;
+  }
+
+  public synchronized void addCommitCallback(final CommitCallback callback) {
+    commitCallback.add(callback);
+  }
+
+  public synchronized void recordConsumedOffsets(
+      final String consumerGroup,
+      final TopicPartition partition,
+      final long offset
+  ) {
+    uncommitted.put(new RecordingKey(partition, consumerGroup), offset);
+  }
+
+  public synchronized void recordProduce(final RecordMetadata recordMetadata) {
+    final TopicPartition partition
+        = new TopicPartition(recordMetadata.topic(), recordMetadata.partition());
+    written.compute(
+        partition,
+        (k, v) -> v == null ? recordMetadata.offset() : Math.max(recordMetadata.offset(), v)
+    );
+  }
+
+  public synchronized void recordCommit() {
+    final Map<RecordingKey, Long> committedOffsets = Map.copyOf(uncommitted);
+    final Map<TopicPartition, Long> writtenOffsets = Map.copyOf(written);
+    commitCallback.forEach(c -> c.onCommit(committedOffsets, writtenOffsets));
+    uncommitted.clear();
+    written.clear();
+  }
+
+  public synchronized void recordAbort() {
+    uncommitted.clear();
+    written.clear();
+  }
+
+  public ResponsiveProducer.Listener getProducerListener() {
+    return producerListener;
+  }
+
+  public ResponsiveConsumer.Listener getConsumerListener() {
+    return consumerListener;
+  }
+
+  @FunctionalInterface
+  interface CommitCallback {
+    void onCommit(
+        final Map<RecordingKey, Long> committedOffsets,
+        final Map<TopicPartition, Long> lastWrittenOffsets
+    );
+  }
+
+  private class ConsumerListener implements ResponsiveConsumer.Listener {
+    @Override
+    public void onCommit(final Map<TopicPartition, OffsetAndMetadata> offsets) {
+      if (eos) {
+        throw new IllegalStateException("consumer commit is not expected with EOS");
+      }
+      offsets.forEach((p, o) -> recordConsumedOffsets("", p, o.offset()));
+      recordCommit();
+    }
+  }
+
+  private class ProducerListener implements ResponsiveProducer.Listener {
+    @Override
+    public void onSendOffsetsToTransaction(Map<TopicPartition, OffsetAndMetadata> offsets,
+                                           String consumerGroupId) {
+      if (!eos) {
+        throw new IllegalStateException("producer commit is not expected with alos");
+      }
+      offsets.forEach((p, o) -> recordConsumedOffsets(consumerGroupId, p, o.offset()));
+    }
+
+    @Override
+    public void onCommit() {
+      // TODO: throw if not eos
```

Comment:
seems easy enough to just do?

Reply:
whoops - I wrote this comment before wiring in the eos flag. Fixing